misc/userscripts/qute-pass

```diff
@@ -0,0 +1,139 @@
+argument_parser.add_argument('--username-target', '-U', choices=['path', 'secret'], default='path',
+                             help='The target for the username regular expression')
+argument_parser.add_argument('--password-pattern', '-P', default=r'(.*)',
+                             help='Regular expression that matches the password')
+argument_parser.add_argument('--dmenu-invocation', '-d', default='rofi -dmenu',
+                             help='Invocation used to execute a dmenu-provider')
+argument_parser.add_argument('--no-insert-mode', '-n', dest='insert_mode', action='store_false',
+                             help="Don't automatically enter insert mode")
+argument_parser.add_argument('--io-encoding', '-i', default='UTF-8',
+                             help='Encoding used to communicate with subprocesses')
+group = argument_parser.add_mutually_exclusive_group()
+group.add_argument('--username-only', '-e', action='store_true', help='Only insert username')
+group.add_argument('--password-only', '-w', action='store_true', help='Only insert password')
+
+stderr = functools.partial(print, file=sys.stderr)
+
+
+class ExitCodes(enum.IntEnum):
+    SUCCESS = 0
+    # 1 is reserved for general script errors
+    NO_PASS_CANDIDATES = 2
+    COULD_NOT_MATCH_USERNAME = 3
+    COULD_NOT_MATCH_PASSWORD = 4
+
+
+def qute_command(command):
+    with open(os.environ['QUTE_FIFO'], 'w') as fifo:
+        fifo.write(command + '\n')
+        fifo.flush()
+
+
+def find_pass_candidates(domain, password_store_path):
+    candidates = []
+    for path, directories, file_names in os.walk(password_store_path):
+        if directories or domain not in path.split(os.path.sep):
+            continue
+
+        # Strip password store path prefix to get the relative pass path
+        pass_path = path[len(password_store_path) + 1:]
+        secrets = fnmatch.filter(file_names, '*.gpg')
+        candidates.extend(os.path.join(pass_path, os.path.splitext(secret)[0]) for secret in secrets)
+    return candidates
+
+
+def pass_(path, encoding):
+    process = subprocess.run(['pass', path], stdout=subprocess.PIPE)
+    return process.stdout.decode(encoding).strip()
+
+
+def dmenu(items, invocation, encoding):
+    command = shlex.split(invocation)
+    process = subprocess.run(command, input='\n'.join(items).encode(encoding), stdout=subprocess.PIPE)
+    return process.stdout.decode(encoding).strip()
+
+
+def main(arguments):
+    domain = tldextract.extract(arguments.url)
+
+    # Expand potential ~ in paths, since this script won't be called from a shell that does it for us
+    password_store_path = os.path.expanduser(arguments.password_store)
+
+    # Try to find candidates using targets in the following order: fully qualified domain (including subdomains),
+    # registered domain and finally the IPv4 address if that's what the URL was
+    for target in filter(None, [domain.fqdn, domain.registered_domain, domain.ipv4]):
```

Comment:
I'd prefer a list comprehension here: `[e for e in [domain.fqdn, domain.registered_domain, domain.ipv4] if e]` - but I'm fine with this too, I'll let you be the judge :wink: